File: app/src/main/java/com/xposed/hook/RimetActivity.java
package com.xposed.hook;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.design.widget.TabLayout;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.telephony.CellLocation;
import android.telephony.PhoneStateListener;
import android.telephony.TelephonyManager;
import android.telephony.gsm.GsmCellLocation;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.xposed.hook.config.Constants;
import com.xposed.hook.config.PkgConfig;
import com.xposed.hook.entity.AppInfo;

public class RimetActivity extends AppCompatActivity implements View.OnClickListener {

    private SharedPreferences sp;

    private TabLayout tabLayout;
    private TabLayout.Tab gpsTab;
    private TabLayout.Tab cellTab;

    private View ll_gps;
    private EditText etLatitude;
    private EditText etLongitude;
    private TextView tvLatitude;
    private TextView tvLongitude;
    private Button btnAutoFillGps;

    private View ll_cell;
    private EditText etLac;
    private EditText etCid;
    private TextView tvLac;
    private TextView tvCid;
    private Button btnAutoFillCell;

    private CompoundButton cb;

    private AppInfo appInfo;
    private boolean isDingTalk;

    TelephonyManager tm;
    GsmCellLocation l;
    LocationManager lm;
    Location gpsL;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        appInfo = (AppInfo) getIntent().getSerializableExtra("appInfo");
        if (appInfo == null)
            return;
        setContentView(R.layout.activity_rimet);
        setTitle(appInfo.title);
        isDingTalk = PkgConfig.pkg_dingtalk.equals(appInfo.packageName);
        tm = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
        lm = (LocationManager) getSystemService(Context.LOCATION_SERVICE);

        String prefix = appInfo.packageName + "_";
        sp = getSharedPreferences(Constants.PREF_FILE_NAME, MODE_WORLD_READABLE);
        String defaultLatitude = Constants.DEFAULT_LATITUDE;
        String defaultLongitude = Constants.DEFAULT_LONGITUDE;
        if (isDingTalk) {
            defaultLatitude = "";
            defaultLongitude = "";
        }

        ll_gps = findViewById(R.id.ll_gps);
        etLatitude = (EditText) findViewById(R.id.et_latitude);
        etLongitude = (EditText) findViewById(R.id.et_longitude);
        tvLatitude = (TextView) findViewById(R.id.tv_latitude);
        tvLongitude = (TextView) findViewById(R.id.tv_longitude);
        btnAutoFillGps = (Button) findViewById(R.id.btn_auto_fill_gps);
        etLatitude.setText(sp.getString(prefix + "latitude", defaultLatitude));
        etLongitude.setText(sp.getString(prefix + "longitude", defaultLongitude));
        btnAutoFillGps.setOnClickListener(this);

        ll_cell = findViewById(R.id.ll_cell);
        etLac = (EditText) findViewById(R.id.et_lac);
        etCid = (EditText) findViewById(R.id.et_cid);
        tvLac = (TextView) findViewById(R.id.tv_lac);
        tvCid = (TextView) findViewById(R.id.tv_cid);
        btnAutoFillCell = (Button) findViewById(R.id.btn_auto_fill_cell);
        int lac = sp.getInt(prefix + "lac", Constants.DEFAULT_LAC);
        int cid = sp.getInt(prefix + "cid", Constants.DEFAULT_CID);
        if (lac != Constants.DEFAULT_LAC)
            etLac.setText(String.valueOf(lac));
        if (cid != Constants.DEFAULT_CID)
            etCid.setText(String.valueOf(cid));
        btnAutoFillCell.setOnClickListener(this);

        initTabLayout();
        cb = (CompoundButton) findViewById(R.id.cb);
        cb.setChecked(sp.getBoolean(appInfo.packageName, false));
        findViewById(R.id.btn_save).setOnClickListener(this);
        findViewById(R.id.btn_reboot_app).setOnClickListener(this);

        requestPermissions();
    }

    private void initTabLayout() {
        tabLayout = (TabLayout) findViewById(R.id.tab_layout);
        gpsTab = tabLayout.newTab().setText(R.string.gps_location);
        cellTab = tabLayout.newTab().setText(R.string.cell_location);
        if (isDingTalk) {
            tabLayout.addTab(cellTab);
            tabLayout.addTab(gpsTab);
            ll_gps.setVisibility(View.GONE);
        } else {
            tabLayout.addTab(gpsTab);
            tabLayout.addTab(cellTab);
            ll_cell.setVisibility(View.GONE);
        }
        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                ll_gps.setVisibility(tab == gpsTab ? View.VISIBLE : View.GONE);
                ll_cell.setVisibility(tab == gpsTab ? View.GONE : View.VISIBLE);
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        switch (id) {
            case R.id.btn_auto_fill_cell:
                etLac.setText(String.valueOf(l.getLac()));
                etCid.setText(String.valueOf(l.getCid()));
                break;
            case R.id.btn_auto_fill_gps:
                etLatitude.setText(String.valueOf(gpsL.getLatitude()));
                etLongitude.setText(String.valueOf(gpsL.getLongitude()));
                break;
            case R.id.btn_save:
                String prefix = appInfo.packageName + "_";
                sp.edit().putString(prefix + "latitude", etLatitude.getText().toString())
                        .putString(prefix + "longitude", etLongitude.getText().toString())
                        .putInt(prefix + "lac", parseInt(etLac.getText().toString()))
                        .putInt(prefix + "cid", parseInt(etCid.getText().toString()))
                        .putBoolean(appInfo.packageName, cb.isChecked())
                        .commit();
                Toast.makeText(getApplicationContext(), R.string.save_success, Toast.LENGTH_SHORT).show();
                break;
            case R.id.btn_reboot_app:
                try {
                    Intent intent = new Intent();
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                    intent.setData(Uri.fromParts("package", appInfo.packageName, null));
                    startActivity(intent);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                break;
        }
    }

    private int parseInt(String str) {
        try {
            return Integer.parseInt(str);
        } catch (Exception e) {
            return -1;
        }
    }

    @Override
    public void finish() {
        stopLocation();
        super.finish();
    }

    PhoneStateListener listener = new PhoneStateListener() {

        @Override
        public void onCellLocationChanged(CellLocation location) {
            if (location instanceof GsmCellLocation) {
                l = (GsmCellLocation) location;
                tvLac.setText(getString(R.string.current_lac, String.valueOf(l.getLac())));
                tvCid.setText(getString(R.string.current_cid, String.valueOf(l.getCid())));
                btnAutoFillCell.setVisibility(View.VISIBLE);
            }
        }
    };

    LocationListener gpsListener = new LocationListener() {
        @Override
        public void onLocationChanged(Location location) {
            gpsL = location;
            tvLatitude.setText(getString(R.string.current_latitude, String.valueOf(location.getLatitude())));
            tvLongitude.setText(getString(R.string.current_longitude, String.valueOf(location.getLongitude())));
            btnAutoFillGps.setVisibility(View.VISIBLE);
        }

        @Override
        public void onStatusChanged(String provider, int status, Bundle extras) {

        }

        @Override
        public void onProviderEnabled(String provider) {

        }

        @Override
        public void onProviderDisabled(String provider) {

        }
    };

    private void requestPermissions() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION}, 101);
            return;
        }
        startLocation();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 101 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startLocation();
        }
    }

    private void startLocation() {
        tm.listen(listener, PhoneStateListener.LISTEN_CELL_LOCATION);
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            return;
        }
        lm.requestSingleUpdate(LocationManager.GPS_PROVIDER, gpsListener, null);
    }

    private void stopLocation() {
        tm.listen(listener, PhoneStateListener.LISTEN_NONE);
        lm.removeUpdates(gpsListener);
    }
}

File: app/src/main/java/com/xposed/hook/RimetActivity.kt
package com.xposed.hook

import android.Manifest
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.telephony.*
import android.telephony.gsm.GsmCellLocation
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.core.app.ActivityCompat
import androidx.lifecycle.MutableLiveData
import com.xposed.hook.config.Constants
import com.xposed.hook.config.PkgConfig
import com.xposed.hook.entity.AppInfo
import com.xposed.hook.theme.AppTheme
import com.xposed.hook.utils.CellLocationHelper
import com.xposed.hook.utils.SharedPreferencesHelper

class RimetActivity : AppCompatActivity() {

    private lateinit var sp: SharedPreferences
    private lateinit var appInfo: AppInfo
    private var isDingTalk = false

    private lateinit var tm: TelephonyManager
    private lateinit var l: GsmCellLocation
    private lateinit var lm: LocationManager
    private lateinit var gpsL: Location

    private val _currentLatitude = MutableLiveData("")
    private val _currentLongitude = MutableLiveData("")
    private val _currentLac = MutableLiveData("")
    private val _currentCid = MutableLiveData("")

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        appInfo = intent.getSerializableExtra("appInfo") as? AppInfo ?: return
        title = appInfo.title
        isDingTalk = PkgConfig.pkg_dingtalk == appInfo.packageName
        tm = getSystemService(TELEPHONY_SERVICE) as TelephonyManager
        lm = getSystemService(LOCATION_SERVICE) as LocationManager

        sp = getSharedPreferences(Constants.PREF_FILE_NAME, MODE_PRIVATE)

        setContent { Container() }

        requestPermissions()
    }

    @Composable
    fun Container() {
        val prefix = appInfo.packageName + "_"
        val defaultLatitude = if (isDingTalk) "" else Constants.DEFAULT_LATITUDE
        val defaultLongitude = if (isDingTalk) "" else Constants.DEFAULT_LONGITUDE
        var latitude by remember {
            mutableStateOf(sp.getString(prefix + "latitude", null) ?: defaultLatitude)
        }
        var longitude by remember {
            mutableStateOf(sp.getString(prefix + "longitude", null) ?: defaultLongitude)
        }
        var lac by remember {
            CellLocationHelper.getLac(sp, prefix).let {
                mutableStateOf(if (it == Constants.DEFAULT_LAC) "" else it.toString())
            }
        }
        var cid by remember {
            CellLocationHelper.getCid(sp, prefix).let {
                mutableStateOf(if (it == Constants.DEFAULT_CID) "" else it.toString())
            }
        }
        var isChecked by remember {
            mutableStateOf(sp.getBoolean(appInfo.packageName, false))
        }
        val currentLatitude by _currentLatitude.observeAsState("")
        val currentLongitude by _currentLongitude.observeAsState("")
        val currentLac by _currentLac.observeAsState("")
        val currentCid by _currentCid.observeAsState("")

        AppTheme {
            Column(
                Modifier
                    .padding(15.dp, 0.dp)
                    .verticalScroll(rememberScrollState())
            ) {
                Text(
                    text = stringResource(R.string.gps_location),
                    modifier = Modifier.padding(0.dp, 16.dp),
                    color = colorResource(R.color.textColorPrimary)
                )
                OutlinedTextField(
                    value = latitude,
                    onValueChange = { latitude = it },
                    label = { Text(text = "latitude") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
                Spacer(modifier = Modifier.height(5.dp))
                OutlinedTextField(
                    value = longitude,
                    onValueChange = { longitude = it },
                    label = { Text(text = "longitude") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
                Spacer(modifier = Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.weight(1f)) {
                        Text(text = stringResource(R.string.current_latitude, currentLatitude))
                        Text(text = stringResource(R.string.current_longitude, currentLongitude))
                    }
                    if (currentLatitude.isNotEmpty())
                        TextButton(onClick = {
                            latitude = currentLatitude
                            longitude = currentLongitude
                        }) {
                            Text(
                                text = stringResource(R.string.auto_fill),
                                color = colorResource(R.color.textColorPrimary)
                            )
                        }
                }
                Text(
                    text = stringResource(R.string.cell_location),
                    modifier = Modifier.padding(0.dp, 16.dp),
                    color = colorResource(R.color.textColorPrimary)
                )
                OutlinedTextField(
                    value = lac,
                    onValueChange = { lac = it },
                    label = { Text(text = "Area Code") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
                Spacer(modifier = Modifier.height(5.dp))
                OutlinedTextField(
                    value = cid,
                    onValueChange = { cid = it },
                    label = { Text(text = "Cell Identity") },
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number)
                )
                Spacer(modifier = Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.weight(1f)) {
                        Text(text = stringResource(R.string.current_lac, currentLac))
                        Text(text = stringResource(R.string.current_cid, currentCid))
                    }
                    if (currentLac.isNotEmpty())
                        TextButton(onClick = {
                            lac = currentLac
                            cid = currentCid
                        }) {
                            Text(
                                text = stringResource(R.string.auto_fill),
                                color = colorResource(R.color.textColorPrimary)
                            )
                        }
                }
                Spacer(modifier = Modifier.height(10.dp))
                Row(
                    modifier = Modifier.height(40.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = stringResource(R.string.open_location_hook),
                        modifier = Modifier.weight(1f)
                    )
                    Switch(checked = isChecked, onCheckedChange = { isChecked = it })
                }
                Row(Modifier.padding(0.dp, 16.dp)) {
                    Button(
                        onClick = {
                            sp.edit().putString(prefix + "latitude", latitude)
                                .putString(prefix + "longitude", longitude)
                                .putLong(prefix + "lac", parseLong(lac))
                                .putLong(prefix + "cid", parseLong(cid))
                                .putLong(prefix + "time", System.currentTimeMillis())
                                .putBoolean(appInfo.packageName, isChecked)
                                .commit()
                            SharedPreferencesHelper.makeWorldReadable(sp)
                            Toast.makeText(
                                applicationContext,
                                R.string.save_success,
                                Toast.LENGTH_SHORT
                            ).show()
                        },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(backgroundColor = Color(0xFF00975C)),
                        contentPadding = PaddingValues(top = 12.dp, bottom = 12.dp)
                    ) {
                        Text(text = stringResource(R.string.save))
                    }
                    Spacer(modifier = Modifier.width(16.dp))
                    Button(
                        onClick = {
                            try {
                                val intent = Intent()
                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                                intent.action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
                                intent.data = Uri.fromParts("package", appInfo.packageName, null)
                                startActivity(intent)
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }
                        },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(backgroundColor = Color(0xFFE9686B)),
                        contentPadding = PaddingValues(top = 12.dp, bottom = 12.dp)
                    ) {
                        Text(text = stringResource(R.string.reboot_app))
                    }
                }
            }
        }
    }

    private fun parseLong(str: String): Long {
        return try {
            str.toLong()
        } catch (e: Exception) {
            -1
        }
    }

    override fun finish() {
        stopLocation()
        super.finish()
    }

    private var listener: PhoneStateListener = object : PhoneStateListener() {
        override fun onCellLocationChanged(location: CellLocation) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) return
            if (location is GsmCellLocation) {
                l = location
                _currentLac.value = l.lac.toString()
                _currentCid.value = l.cid.toString()
            }
        }

        override fun onCellInfoChanged(cellInfo: MutableList<CellInfo>?) {
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) return
            if (cellInfo == null || cellInfo.isEmpty()) return
            when (val cellIdentity = cellInfo[0].cellIdentity) {
                is CellIdentityGsm -> {
                    _currentLac.value = cellIdentity.lac.toString()
                    _currentCid.value = cellIdentity.cid.toString()
                }
                is CellIdentityLte -> {
                    _currentLac.value = cellIdentity.tac.toString()
                    _currentCid.value = cellIdentity.ci.toString()
                }
                is CellIdentityNr -> {
                    _currentLac.value = cellIdentity.tac.toString()
                    _currentCid.value = cellIdentity.nci.toString()
                }
            }
        }
    }

    private var gpsListener: LocationListener = object : LocationListener {
        override fun onLocationChanged(location: Location) {
            gpsL = location
            _currentLatitude.value = location.latitude.toString()
            _currentLongitude.value = location.longitude.toString()
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}
        override fun onProviderEnabled(provider: String) {}
        override fun onProviderDisabled(provider: String) {}
    }

    private fun requestPermissions() {
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this, arrayOf(
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION
                ), 101
            )
            return
        }
        startLocation()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 101 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startLocation()
        }
    }

    private fun startLocation() {
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
            && ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        lm.requestSingleUpdate(LocationManager.GPS_PROVIDER, gpsListener, null)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
            tm.listen(listener, PhoneStateListener.LISTEN_CELL_LOCATION)
        } else {
            tm.listen(listener, PhoneStateListener.LISTEN_CELL_INFO)
        }
    }

    private fun stopLocation() {
        tm.listen(listener, PhoneStateListener.LISTEN_NONE)
        lm.removeUpdates(gpsListener)
    }
}